toolkit/tools/imagegen/diskutils/filesystem.go

```diff
@@ -0,0 +1,316 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package diskutils
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/microsoft/azurelinux/toolkit/tools/internal/kernelversion"
+	"github.com/microsoft/azurelinux/toolkit/tools/internal/logger"
+	"github.com/microsoft/azurelinux/toolkit/tools/internal/shell"
+	"github.com/microsoft/azurelinux/toolkit/tools/internal/sliceutils"
+	"github.com/microsoft/azurelinux/toolkit/tools/internal/targetos"
+	"github.com/microsoft/azurelinux/toolkit/tools/internal/version"
+)
+
+// When calling mkfs, the default options change depending on the host OS you are running on and typically match
+// what the distro has decided is best for their OS. For example, for ext2/3/4, the defaults are stored in
+// /etc/mke2fs.conf.
+// However, when building Azure Linux images, the defaults should be as consistent as possible and should only contain
+// features that are supported on Azure Linux.
+
+type ext4Options struct {
+	BlockSize int
+	Features  []string
+}
+
+type xfsOptions struct {
+	Features []string
+}
+
+type fileSystemsOptions struct {
+	Ext4 ext4Options
+	Xfs  xfsOptions
+}
+
+var (
+	azl2Ext4Options = ext4Options{
+		BlockSize: 4096,
+		Features: []string{"sparse_super", "large_file", "filetype", "resize_inode", "dir_index", "ext_attr",
+			"has_journal", "extent", "huge_file", "flex_bg", "metadata_csum", "64bit", "dir_nlink", "extra_isize",
+		},
+	}
+
+	azl3Ext4Options = ext4Options{
+		BlockSize: 4096,
+		Features: []string{"sparse_super", "large_file", "filetype", "resize_inode", "dir_index", "ext_attr",
+			"has_journal", "extent", "huge_file", "flex_bg", "metadata_csum", "64bit", "dir_nlink", "extra_isize",
+			"orphan_file",
```

Comment:
Question, why was `orphan_file` added to Azl3? Perhaps adding a clue comment explaining this diff between Azl2 and Azl3 for ext4.

Reply:
Fixed.